utils/inference_utils.py

```diff
@@ -844,19 +844,29 @@ def run_inference(examples, serving_bundle):
     return (common_utils.convert_prediction_values(values, serving_bundle),
             None)
   elif serving_bundle.custom_predict_fn:
-    # If custom_predict_fn is provided, pass examples directly for local
-    # inference.
-    values = serving_bundle.custom_predict_fn(examples)
-    extra_results = None
-    # If the custom prediction function returned a dict, then parse out the
-    # prediction scores. If it is just a list, then the results are the
-    # prediction results without attributions or other data.
-    if isinstance(values, dict):
-      preds = values.pop('predictions')
-      extra_results = values
-    else:
-      preds = values
-    return (common_utils.convert_prediction_values(preds, serving_bundle),
-            extra_results)
+    from inspect import signature
+    sig = signature(serving_bundle.custom_predict_fn)
+    params = sig.parameters
+    # The custom_predict_fn for colab/jupyter accepts one parameter.
+    # While the custom_predict_fn for non-colab usage have two.
+    if len(params) == 1:
+      # If custom_predict_fn is provided, pass examples directly for local
+      # inference.
+      values = serving_bundle.custom_predict_fn(examples)
+      extra_results = None
+      # If the custom prediction function returned a dict, then parse out the
+      # prediction scores. If it is just a list, then the results are the
+      # prediction results without attributions or other data.
+      if isinstance(values, dict):
+        preds = values.pop('predictions')
+        extra_results = values
+      else:
+        preds = values
+      return (common_utils.convert_prediction_values(preds, serving_bundle),
+              extra_results)
+
+    if len(params) == 2:
+      return (serving_bundle.custom_predict_fn(examples, serving_bundle), None)
```

Comment:
for simplicity, have this function have the same return semantics as the normal custom_predict_fn case, and have it call lines 856-866 in both cases. then when you test it, have your custom predict fn just return a list of lists of classification scores, instead of having it return the protocol buffers. the convert_prediction_values method will convert those raw lists into the protos for you.

Reply:
Totally agreed! Fixed and tested with classification and regression. But I can't color the points by its "class label" in the front-end now.